<a href="https://colab.research.google.com/github/ReneSRuiz/Mineria_de_datos/blob/main/Avance1_PIA_Equipo_11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Equipo:** 11  
**Grupo:** 001  
**Frecuencia:** lunes  
**Integrantes:**

 *   Alejandro Uriel García Alanís 1886968
 *   Jorge Eduardo Estrada Dávila 1741943
 *   Néstor Misael Paz Reyes 1559508
 *   Elisa González García 1858207
 *   Rene Sobrevilla Ruiz 1941452 


**BASE DE DATOS PRINCIPAL**  
**1) Título de la base de datos**  
En nombre oficial de la base de datos es "Lichess September 2020 data", la cual se puede encontrar por el siguiente enlace: https://www.kaggle.com/noobiedatascientist/lichess-september-2020-data?select=Sept_20_analysis.csv

**2) Introducción general**  
El ajedrez es un juego con muchos años de antigüedad, sencillo de entender pero con una dificultad estratégica muy interesante. Se cree que los grandes reyes y estrategas militares veían al ajedrez como una forma de estudiar a sus enemigos y mejorar como líderes. El ajedrez a tenido momentos importantes en la historia humana, desde juegos amistosos para generar alianzas entre reyes como el fin de épocas importantes como lo fué la primera victoria de un americano contra la URSS, país que dominaba este deporte en tiempos de guerra fría, significando una victoria moral para EUA.  
El ajedrez era un deporte de nicho, popular entre gente que buscaba un juego estratégico y gente que ama el análisis tan complejo de este. Pero en tiempos modernos, con estrenos de series muy famosas, ejemplo Gambito de Dama, este juego ha alcanzado una popularuidad muy alta. Ahora, no solamente las personas "inteligentes" lo juegan, sino que la persona común se ha animado a aprender de este juego tan impresionante y educativo.

**3) Descripción de los datos**  
Los datos de la base seleccionada estan en una tabla. En total, son más de 3 millones de filas, y 40 columnas. Cada fila representa un enfrentamiento entre 2 personas y las columnas nos revelan varios datos, como el tipo de juego, el "elo" de los jugadores, la fecha del juego, resultados, momentos importantes, nombre de aperturas, movimientos y errores de ambos, entre otros. Hay datos tanto enteros, flotantes y objetos (cadenas de caracteres y fechas).

**4) Justificación del uso de datos.**   
Creemos que la base de datos es muy completa, ya que tiene detallado cada movimiento de los jugadores, como el estado de la partida que estos jugaron. Además, la información está muy completa, sin muchos datos nulos que puedan afectar nuestro trabajo.
Hablando de los beneficios, al estar en su mayoria completos, creemos que podemos obtener resultados significativos y que nos ayuden a solucionar el problema planteado y analizar este juego de una manera más significativa. Es verdad que al ser muchos datos, se tardará en realizar los procesos programados, pero con paciencia, estos pueden ser muy reveladores.

**5) Objetivo final.**  
Nuestro objetivo principal es poder clasificar las mejores aperturas del ajedrez, dependiento del "elo" que tengan los jugadores y la eficacia de estas aperturas para la creación de entrategias ganadoras.  
Creemos que, con estos datos, podemos crear la clasificación antes mencionada y ayudar a las personas nuevas, o un poco mas experimentadas, a saber cual apertura funciona mejor dependiendo de su desempeño como de gustos, y asi, poder estudiar la apertura perfecta o cambiar su estrategia para que esta se ajuste a su estilo.

**6) Mencionar preguntas de interés.**  
Hay varias cosas interesantes que podemos indagar sobre esta base, como lo pueden ser:  
* ¿Se juegan más aperturas clásicas o modernas?  
* ¿El tiempo promedio de juego aumenta conforme el rival?, es decir, ¿Mientras mas "elo" tienen los rivales, mas largas son estas partidas?
* ¿Son igual de efectivas las aperturas populares entre la gente de "elo bajo" en los "elos altos"?  
* ¿Las aperturas favoritas de los "elos altos" son efectivas en los "elos bajos"?
* ¿Que tantas jugadas inadecuadas, meteduiras de patas o volteos de juego tiene un jugador de "low elo" comparado con los de mayor "elo"?

**7) Parte de programación.**  
* **Importar librerias y base de datos.**  

In [ ]:
#importando las librerias
import pandas as pd
import numpy as np
#Importar la base de datos al google colab
from google.colab import files
files.upload()
#Nuestros datos a trabajar son
df = pd.read_csv("Sept_20_analysis.csv")

Saving Sept_20_analysis.csv to Sept_20_analysis.csv


* **Visualizar los datos.**  
Aquí veremos los primeros 3 datos, los últimos 3 datos y al final, los nombres de las columnas.

In [ ]:
#Veamos los primeros 3 datos
print(df.head(3), "\n")
#Veamos los últimos 3 datos
print(df.tail(3),"\n")
#Veamos los nombres de las columnas
print("\n", df.columns, "\n")

   GAME  BlackElo  ...  Game_flips Game_flips_ts
0    11      1143  ...           8             0
1    14      1504  ...           6             0
2    29      1933  ...           5             0

[3 rows x 40 columns] 

          GAME  BlackElo  ...  Game_flips Game_flips_ts
3739906  65499      1160  ...          13             0
3739907  65544      1306  ...           2             0
3739908  65549      1480  ...           7             0

[3 rows x 40 columns] 


 Index(['GAME', 'BlackElo', 'BlackRatingDiff', 'Date', 'ECO', 'Event',
       'Opening', 'Result', 'Site', 'Termination', 'TimeControl', 'UTCTime',
       'WhiteElo', 'WhiteRatingDiff', 'Black_elo_category',
       'White_elo_category', 'starting_time', 'increment', 'Game_type',
       'Total_moves', 'Black_blunders', 'White_blunders', 'Black_mistakes',
       'White_mistakes', 'Black_inaccuracies', 'White_inaccuracies',
       'Black_inferior_moves', 'White_inferior_moves', 'Black_ts_moves',
       'White_ts_moves', 'Black

* **Descrpcion básica de columnas (tipo, elementos, celdas vacias, etc.)**  

In [ ]:
#Veamos las dimensiones de nuestro dataframe
print("Dimensiones: ", df.ndim)
#Veamos el número de filas
print("Filas: ",df.shape[0])
#Veamos el número de columnas
print("Columnas: ",df.shape[1])
#Veamos la cantidad de datos de nuestro dataframe
print("Datos totales: ", df.size, "\n")
#Veamos el tipo de datos por columna
print(df.dtypes, "\n")
#Vemos si hay datos vacios
df.isnull().sum()

Dimensiones:  2
Filas:  3739909
Columnas:  40
Datos totales:  149596360 

GAME                      int64
BlackElo                  int64
BlackRatingDiff         float64
Date                     object
ECO                      object
Event                    object
Opening                  object
Result                   object
Site                     object
Termination              object
TimeControl              object
UTCTime                  object
WhiteElo                  int64
WhiteRatingDiff         float64
Black_elo_category       object
White_elo_category       object
starting_time           float64
increment               float64
Game_type                object
Total_moves               int64
Black_blunders            int64
White_blunders            int64
Black_mistakes            int64
White_mistakes            int64
Black_inaccuracies        int64
White_inaccuracies        int64
Black_inferior_moves      int64
White_inferior_moves      int64
Black_ts_moves            int6

GAME                        0
BlackElo                    0
BlackRatingDiff         13036
Date                        0
ECO                         0
Event                       0
Opening                     0
Result                      0
Site                        0
Termination                 0
TimeControl                 0
UTCTime                     0
WhiteElo                    0
WhiteRatingDiff         13057
Black_elo_category          0
White_elo_category          0
starting_time           10344
increment               10344
Game_type                   0
Total_moves                 0
Black_blunders              0
White_blunders              0
Black_mistakes              0
White_mistakes              0
Black_inaccuracies          0
White_inaccuracies          0
Black_inferior_moves        0
White_inferior_moves        0
Black_ts_moves              0
White_ts_moves              0
Black_ts_blunders           0
White_ts_blunders           0
Black_ts_mistakes           0
White_ts_m

Vemos que la base de datos es de 2 dimensiones, lo cual era lo esperado. Esta base cuenta con 3739909 filas, las cuales representan los juegos a analisar, y 40 columnas. En total tenemos 149596360 datos. Vemos que tenemos tanto datos enteros, como objetos y unos pocos flotantes. Hablando de los datos nulos, solamente tenemos 4 columnas con datos nulos:
* BlackRatingDiff, que es el cambio de "elo" del jugador de negras.  
* WhiteRatingDiff, que es el cambio de "elo" del jugador de blancas.  
* starting_time, el tiempo que tienen los jugadores al inicio.  
* increment, el incremento del reloj.  

Estas columnas serian candidatas a ser eliminadas o no ser tomadas en cuenta.


* **Cambiar nombres de columnas.**  

Como todos son en inglés, cambiaremos los nombres a español, pero solamente de las columnas de interés. Primeramente eliminaremos las columnas que pensamos no aportan tanto valor y después haremos el cambio de nombre.

In [ ]:
#Eliminación de columnas
df_temp=df.drop({'GAME','Event','BlackRatingDiff','Date','ECO','Site','UTCTime','WhiteRatingDiff','starting_time','increment',
                 'Black_inferior_moves', 'White_inferior_moves', 'Black_ts_moves','White_ts_moves','Black_ts_blunders', 
                 'White_ts_blunders','Black_ts_mistakes', 'White_ts_mistake', 'Black_long_moves','White_long_moves',
                 'Black_bad_long_moves', 'White_bad_long_moves','Game_flips_ts'},axis = 1 )
#Cambiaremos los nombres de inglés a español
df_temp=df_temp.rename(columns = {'Opening':'Apertura','Result':'Resultado','Termination':'Terminacion',
                                    'TimeControl':'ControlTiempo','Black_elo_category':'Categoria_Negras',
                                    'White_elo_category':'Categoria_Blancas','Game_type':'Tipo_juego',
                                    'Total_moves':'Movimientos','Black_blunders':'Negras_Metedura_de_patas',
                                    'White_blunders':'Blancas_Metedura_de_patas','Black_mistakes':'Negras_Errores',
                                    'White_mistakes':'Blancas_Errores','Black_inaccuracies':'Negras_Imprecision',
                                    'White_inaccuracies':'Blancas_Imprecision','Game_flips':'Juego_Vueltas',})
df_temp.head(3)

BlackElo          Apertura  ... Blancas_Imprecision Juego_Vueltas
0      1143      Bird Opening  ...                   1             8
1      1504      Réti Opening  ...                   5             6
2      1933  Philidor Defense  ...                   8             5

[3 rows x 17 columns]

In [ ]:
df_temp.columns

Index(['BlackElo', 'Apertura', 'Resultado', 'Terminacion', 'ControlTiempo',
       'WhiteElo', 'Categoria_Negras', 'Categoria_Blancas', 'Tipo_juego',
       'Movimientos', 'Negras_Metedura_de_patas', 'Blancas_Metedura_de_patas',
       'Negras_Errores', 'Blancas_Errores', 'Negras_Imprecision',
       'Blancas_Imprecision', 'Juego_Vueltas'],
      dtype='object')

* **Clasificación de los datos**  

Veremos los valores únicos de las categorias, y con estos, crear la clasificación por tipo de jugador dependiendo de la categoría que salga. Como nos enfocamos en la s aperturas, usaremos la categoría de las blancas para esta clasificación.

In [ ]:
#En este caso, vamos a clasificar los datos por la categoría de los jugadores, entonces veamos si hay valores únicos:
Categorias=df_temp.Categoria_Blancas.unique()
print(Categorias, "\n")
#Ahora hagamos un df por categoría
Low_rating = df_temp[df_temp['Categoria_Blancas']==Categorias[0]]
High_rating = df_temp[df_temp['Categoria_Blancas']==Categorias[1]]
GM_rating = df_temp[df_temp['Categoria_Blancas']==Categorias[2]]
#Comprobemos si se realizaron los cambios:
print(Low_rating.head(3), "\n")
print(High_rating.head(3), "\n")
print(GM_rating.head(3), "\n")

['Low rating' 'High rating' 'GM rating'] 

   BlackElo          Apertura  ... Blancas_Imprecision Juego_Vueltas
0      1143      Bird Opening  ...                   1             8
1      1504      Réti Opening  ...                   5             6
2      1933  Philidor Defense  ...                   8             5

[3 rows x 17 columns] 

   BlackElo  ... Juego_Vueltas
3      1710  ...             8
4      1598  ...             2
5      2207  ...             6

[3 rows x 17 columns] 

     BlackElo  ... Juego_Vueltas
48       2300  ...             6
112      2362  ...             3
360      2361  ...             3

[3 rows x 17 columns] 



* **Guardar otros archivos con agrupaciones.**  

In [ ]:
#Crearemos 3 archivos distintos por la agrupación que hagamos
Low_rating.to_csv("Low_rating.csv", index=False)
High_rating.to_csv("High_rating.csv", index=False)
GM_rating.to_csv("GM_rating.csv", index=False)

**BASE DE DATOS SECUNDARIA**  
**1) Título de la base de datos**  
En nombre oficial de la base de datos es "Students Performance in Exams", la cual se puede encontrar por el siguiente enlace: https://www.kaggle.com/spscientist/students-performance-in-exams

**2) Introducción general**  
La escuela, como sabemos, es una etapa primordial en el crecimiento de una persona, pero a su ves, tiene momentos de angustia. Uno de esos momentos son los exámenes, ya que el aplitar todo lo aprendido en ejercicios/preguntas con ñímite de tiempo, puede generar estres en los estudiantes y fallar. Por ello, siempre el estudio, los exámenes de preparación e incluso la familia juega un papel importante a la hora de presentar estos.

**3) Descripción de los datos**  
Los datos muerstran, a grandes rasgos, los resultados de 3 exámenes realizados a diferentes grupos (A, B, C, D). Estos se encuentran en una tabla de 1000 filas y 8 columnas, las cuales son las siguientes:
* gender, el sexo del estudiante.
* race/ethnicity, el grupo al que pertenece.
* parental level of education, grado de educación de los padres.
* lunch, tipo de recreo.
* test preparation course, si timó o no el examen de preparación.
* math score, nota del examen de matemáticas.
* reading score, nota del examen de lectura.
* writing score, nota del examen de escritura.

**4) Justificación del uso de datos.**  
Creemos que es una buena base de datos por el resumen bien hecho que tiene la misma. Creemos que, si se usa esta base de datos, podremos obtener resultados significativos. Además que es facil de manipular y visualizar los datos por la forma en como está construida.

**5) Objetivo final.**  
El objetivo final es encontrar si existe relación alguna el tomar el examen de preparacion y nivel educativo de los padres con respecto a las calificaciones obtenias en el examnes. Con esto, se podrian crear estrategias en las escuelas que ayuden a preparar a sus estudiantes de una manera más interesante.

**6) Mencionar preguntas de interés.**  
Hay varias cosas interesantes que podemos indagar sobre esta base, como lo pueden ser:  
* ¿Los hijos con padres "estudiados" no necesitan de exámenes de preparación para sacar buenas calificaciones?
* ¿El venir de padres sin tantos "estudios" condena al estudiante a esforzarse más?
* ¿Realmente son efectivos los cursos de preparación?

**7) Parte de programación.**  
* **Importar librerias y base de datos.**  

In [ ]:
#importando las librerias
import pandas as pd
import numpy as np
#Importar la base de datos al google colab
from google.colab import files
files.upload()
#Nuestros datos a trabajar son
df2 = pd.read_csv("StudentsPerformance.csv")

Saving StudentsPerformance.csv to StudentsPerformance.csv


* **Visualizar los datos.**  
Aquí veremos los primeros 3 datos, los últimos 3 datos y al final, los nombres de las columnas

In [ ]:
#Veamos los primeros 3 datos
print(df2.head(3), "\n")
#Veamos los últimos 3 datos
print(df2.tail(3),"\n")
#Veamos los nombres de las columnas
print("\n", df2.columns, "\n")

   gender race/ethnicity  ... reading score writing score
0  female        group B  ...            72            74
1  female        group C  ...            90            88
2  female        group B  ...            95            93

[3 rows x 8 columns] 

     gender race/ethnicity  ... reading score writing score
997  female        group C  ...            71            65
998  female        group D  ...            78            77
999  female        group D  ...            86            86

[3 rows x 8 columns] 


 Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object') 



* **Descrpcion básica de columnas (tipo, elementos, celdas vacias, etc.)**

In [ ]:
#Veamos las dimensiones de nuestro dataframe
print("Dimensiones: ", df2.ndim)
#Veamos el número de filas
print("Filas: ",df2.shape[0])
#Veamos el número de columnas
print("Columnas: ",df2.shape[1])
#Veamos la cantidad de datos de nuestro dataframe
print("Datos totales: ", df2.size, "\n")
#Veamos el tipo de datos por columna
print(df2.dtypes, "\n")
#Vemos si hay datos vacios
df2.isnull().sum()

Dimensiones:  2
Filas:  1000
Columnas:  8
Datos totales:  8000 

gender                         object
race/ethnicity                 object
parental level of education    object
lunch                          object
test preparation course        object
math score                      int64
reading score                   int64
writing score                   int64
dtype: object 



gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

Vemos que la base de datos es de 2 dimensiones, lo cual era lo esperado. Esta base cuenta con 1000 filas, las cuales son información es 1 estudiante por fila, y 8 columnas. En total tenemos 8000 datos. Vemos que solamente hay datos enteros como objetos, que en este caso son cadenas de caracteres. Hablando de los datos nulos, no hay ninguno en esta base de datos, lo cual es buena señal.

* **Cambiar nombres de columnas.**  

Como todos son en inglés, cambiaremos los nombres a español, pero solamente de las columnas de interés. Primeramente eliminaremos las columnas que pensamos no aportan tanto valor y después haremos el cambio de nombre.

In [ ]:
#Eliminación de columnas
df2_temp=df2.drop('race/ethnicity',axis = 1 )
#Cambiaremos los nombres de inglés a español
df2_temp=df2_temp.rename(columns = {'gender':'Sexo','parental level of education':'Nivel_educativo_padres',
                                    'lunch':'Recreo','test preparation course':'Examen_preparacion',
                                    'math score':'Examen_matematicas','reading score':'Examen_lectura',
                                    'writing score':'Examen_escritura'})
df2_temp.head(3)

Sexo Nivel_educativo_padres  ... Examen_lectura Examen_escritura
0  female      bachelor's degree  ...             72               74
1  female           some college  ...             90               88
2  female        master's degree  ...             95               93

[3 rows x 7 columns]

In [ ]:
df2_temp.columns

Index(['Sexo', 'Nivel_educativo_padres', 'Recreo', 'Examen_preparacion',
       'Examen_matematicas', 'Examen_lectura', 'Examen_escritura'],
      dtype='object')

* **Clasificación de los datos**  

Haremos una pequeña clasificación por tipo de nivel educativo de los padres.

In [ ]:
#En este caso, vamos a clasificar los datos por nivel educativo de los padres, entonces veamos si hay valores únicos:
N_Padres=df2_temp.Nivel_educativo_padres.unique()
print(N_Padres, "\n")
#Ahora hagamos un df por categoría
bachelors_degree = df2_temp[df2_temp['Nivel_educativo_padres']==N_Padres[0]]
some_college = df2_temp[df2_temp['Nivel_educativo_padres']==N_Padres[1]]
masters_degree = df2_temp[df2_temp['Nivel_educativo_padres']==N_Padres[2]]
associates_degree = df2_temp[df2_temp['Nivel_educativo_padres']==N_Padres[3]]
high_school = df2_temp[df2_temp['Nivel_educativo_padres']==N_Padres[4]]
some_high_school = df2_temp[df2_temp['Nivel_educativo_padres']==N_Padres[5]]
#Comprobemos si se realizaron los cambios:
print(bachelors_degree.head(3), "\n")
print(some_college.head(3), "\n")
print(masters_degree.head(3), "\n")
print(associates_degree.head(3), "\n")
print(high_school.head(3), "\n")
print(some_high_school.head(3), "\n")

["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school'] 

      Sexo Nivel_educativo_padres  ... Examen_lectura Examen_escritura
0   female      bachelor's degree  ...             72               74
24    male      bachelor's degree  ...             71               80
27  female      bachelor's degree  ...             69               75

[3 rows x 7 columns] 

     Sexo Nivel_educativo_padres  ... Examen_lectura Examen_escritura
1  female           some college  ...             90               88
4    male           some college  ...             78               75
6  female           some college  ...             95               92

[3 rows x 7 columns] 

      Sexo Nivel_educativo_padres  ... Examen_lectura Examen_escritura
2   female        master's degree  ...             95               93
14  female        master's degree  ...             53               58
18    male        master's degree  ...             42          

* **Guardar otros archivos con agrupaciones.**  

In [ ]:
#Crearemos 6 archivos distintos por la agrupación que hagamos
bachelors_degree.to_csv("bachelors_degree.csv", index=False)
some_college.to_csv("some_college.csv", index=False)
masters_degree.to_csv("masters_degree.csv", index=False)
associates_degree.to_csv("associates_degree.csv", index=False)
high_school.to_csv("high_school.csv", index=False)
some_high_school.to_csv("some_high_school.csv", index=False)